src/mono/mono/metadata/metadata.c

```diff
@@ -7077,23 +7036,24 @@ mono_type_to_unmanaged (MonoType *type, MonoMarshalSpec *mspec, gboolean as_fiel
 const char*
 mono_metadata_get_marshal_info (MonoImage *meta, guint32 idx, gboolean is_field)
 {
-	locator_t loc = {0,};
-	MonoTableInfo *tdef  = &meta->tables [MONO_TABLE_FIELDMARSHAL];
-
-	loc.t = tdef;
-	loc.col_idx = MONO_FIELD_MARSHAL_PARENT;
-	loc.idx = ((idx + 1) << MONO_HAS_FIELD_MARSHAL_BITS) | (is_field? MONO_HAS_FIELD_MARSHAL_FIELDSREF: MONO_HAS_FIELD_MARSHAL_PARAMDEF);
-
-	/* FIXME: Index translation */
-
-	gboolean found = tdef->base && mono_binary_search (&loc, tdef->base, table_info_get_rows (tdef), tdef->row_size, table_locator);
-
-        if (G_UNLIKELY (meta->has_updates)) {
-                if (!found && !mono_metadata_update_metadata_linear_search (meta, tdef, &loc, table_locator))
-                        return NULL;
-        }
+	guint32 tok = mono_metadata_make_token((is_field ? MONO_TABLE_FIELD: MONO_TABLE_PARAM), idx);
+	mdcursor_t marshal_info;
+	uint32_t count;
+	if (!md_create_cursor (meta->metadata_handle, mdtid_FieldMarshal, &marshal_info, &count))
+		return NULL;
+	
+	if (!md_find_row_from_cursor (marshal_info, mdtFieldMarshal_Parent, tok, &marshal_info))
+		return NULL;
 
-	return mono_metadata_blob_heap (meta, mono_metadata_decode_row_col (tdef, loc.result, MONO_FIELD_MARSHAL_NATIVE_TYPE));
+	// TODO-DNMD: Hot Reload.
+	// If we go with DNMD's metadata update mechanism, then we don't need to do anything special here.
+	// If we keep Mono's layered mechanism and don't use DNMD's md_apply_delta API, then we'll need to do something here.
```

Comment:
For the most part, mono's layered approach didn't work out:  I was hoping to avoid having to fully materialize the merged metadata tables after an update (ie: make row decoding slower but avoid having to place mutated image data into writable memory).  But eventually I ended up giving up on that so now Mono ends up creating mutated copies of every modified table when applying a delta anyway.

There is one artifact of the original design that's kind of subtle: we never have to stop the world while we're applying the update because we still never write over old table data - we always create copies. So when we use the managed hot reload API, other threads keep running and using the old tables until the update is done.

I think if there was in-place mutation (I assume that's what DNMD does - AFAICT that's what CoreCLR's EnC does), we would have to be careful to stop the world before applying the update (easy, the GC's STW api is available).  And ideally we'd have some way to ensure that no thread is inside the interpreter's transform stage when the world stops - we don't want it to see a mix of old and new metadata.